### ISSUE TITLE
Minimum date on Date component does not validate typed input

### ISSUE THREAD

**Issue by patrickmelvin** [Sage/carbon#1749]
Inputting a date manually without the date picker will allow a user to skip validation on the minimum date.

**Comment by Euklidian-Space** [comment]
Hi @patrickmelvin .  I would like to tackle this issue.  I have a couple questions though.  I think the best approach would be to have an inequality check inside handleVisibleInputChange method in side the Date component.  There we can compare the date in the input component with the minDate property of the Date component.  However, it would probably be prudent to notify the user by using the input validation decorator.  My question is how is that done exactly?  How an input field is invalidated is kind of lost to me.  It is hard to follow how the various input decorators are composed to accomplish the invalid input field highlight.  Any insight on the matter would be great.  

Thanks,
Gabriel

**Comment by andrewjtait** [comment]
Hi @Euklidian-Space - thanks for taking a look.

I think you are correct in that we should make use of the validation API present on the input components. The easiest way to do that would be to create a new validator, or adapt the current [DateValidator](https://github.com/Sage/carbon/blob/master/src/utils/validations/date/date.js) to include options for min and max date.

This way you could just feed these props into the pre-applied validation here - https://github.com/Sage/carbon/blob/master/src/components/date/date.js#L132

For example:

```js
internalValidations: [new DateValidator({
  minDate: this.props.minDate,
  maxDate: this.props.maxDate
})]
```

The validation decorator would take care of the rest.

I agree the decorators are a bit too abstract to make much sense of, they are something we would love to refactor in the future to instead make use of higher order components or something.

**Comment by Euklidian-Space** [comment]
I think making changes to the DateValidator is the safer bet.  I am going to start on this now.  Should have a PR soon.

**Comment by Euklidian-Space** [comment]
@andrewjtait A few problems arising.  The "pre-applied" validation you spoke is a member of defaultProps which is static.  We cannot reference "this" in there.  From the looks of it, the method _validations in the input-validations.js decorator concatenates this.props.validations with with the internalValidations array in default props.  However, where "this.props.validations" is created or mutated is unclear to me.  Any thoughts?

Thanks,
Gabriel

**Comment by andrewjtait** [comment]
Having a think about it, it makes sense that you cannot access `this.props` in the static method, as they may change during the lifetime of the instance.

Looking into this code makes me more interested to rewrite it!

What if `internalValidations` could be a function which returns an array? This way it is evaluated at the point it requires the validations, and has access to the latest props - even if they have changed since the instance was created.

For example:

```
internalValidations: props => {
  return [new DateValidator({
    minDate: props.minDate,
    maxDate: props.maxDate
  })];
}
```

This would of course require a minor change to `input-validations.js` here: https://github.com/Sage/carbon/blob/master/src/utils/decorators/input-validation/input-validation.js#L731

Which would do a `typeof` check on `internalValidations`. I think this way we ensure the validations are using the latest props, and don't have to worry about re-creating the validator instance if they change etc, and we don't have to track the validations anywhere.

Also the `DateRangeValidator` is used for multiple Date components, such as a "start date" and "end date", and checks the difference between the two - I think it may have a different purpose than what you're trying to achieve here.

**Comment by Euklidian-Space** [comment]
@andrewjtait I like your idea for inputValidations.  I see your point about the DateRangeValidator, however, there probably will be duplicate code in the DateValidator.  The logic we will need DateValidator to have, to solve this issue, will contain the logic of DateRangeValidator since we essentially have a date range with minDat and maxDate.

**Comment by Euklidian-Space** [comment]
@andrewjtait .  I like that idea.  I will start working on it.  I'll have a PR soon.

**Comment by Euklidian-Space** [comment]
@andrewjtait Running into some issues.  I sort of got things working the way we discussed.  When a minDate or maxDate are passed as props the input field will show errors if the date entered in the input is out of the range of minDate and maxDate.  However, for some reason minDate and maxDate assumes the dd/mm/yyyy format.  This causes a discrepancy in what you can type in the input field and what is selectable on the date picker.  It also causes unpredictable results if someone manually enters a date in a different format, say mm/dd/yyyy.  My code that checks if a date is in the range is located in my changes to DateHelper.  

```javascript
  withinDateRange: (value, startDate, endDate) => {
    let [parsedValue, parsedStartDate, parsedEndDate] = [value, startDate, endDate]
      .map(date => DateHelper._parseDate(date));
    if (!parsedValue.isValid()) {
      return true;
    } else if (parsedStartDate.isValid() && parsedEndDate.isValid()) {
      parsedStartDate = parsedStartDate.unix();
      parsedEndDate = parsedEndDate.unix();
    } else if (parsedStartDate.isValid()) {
      parsedStartDate = parsedStartDate.unix();
      parsedEndDate = Infinity;
    } else if (parsedEndDate.isValid()) {
      parsedStartDate = -Infinity;
      parsedEndDate = parsedEndDate.unix();
    } else {
      return true;
    }
    parsedValue = parsedValue.unix();
    
    return parsedStartDate <= parsedValue && parsedEndDate >= parsedValue; 
  },
```

As you can see I am using DateHelper._parseDate() to convert the strings value, startDate, and endDate in to moment objects and then convert those dates to unix time to do the comparison.  For some reason this gives erroneous results when value is not dd/mm/yyyy format, and I am not sure why.  Any ideas?

**Comment by andrewjtait** [comment]
Hi @Euklidian-Space 

We normally work with the ISO format internally (`YYYY-MM-DD`). You're probably seeing the error because you're relying on a localised version, which by default in our demo site is UK format (`DD/MM/YYYY`). I would try switching to the ISO format for the min and max values, we only use the localised version to display the value to the user.

Looking at your code, you could maybe simplify it a bit with 2 changes:

* Lift the validation of the min/max date values into the constructor (it could maybe raise a JS error or warning if they're not valid).
* Use moment.js's methods of date comparison, rather than converting into a unix value - https://momentjs.com/docs/#/query/

**Comment by Euklidian-Space** [comment]
Hey @andrewjtait.  Sorry, but I have been super busy and I just now have some time to put towards this issue.  I have a working solution for this issue.  However I am unclear on how I should go about writing tests for it.  I have unit tests for the peripheral changes I had to make, but testing this issue I am not too sure.  I was thinking about maybe creating a jest Mock for the internalValidations property of the Date component, but I am not clear on how I can access that property in the test environment.

**Comment by andrewjtait** [comment]
Hi @Euklidian-Space - if you have some work in progress it'd be great to see a PR in progress and we can advise on there?

**Comment by Euklidian-Space** [comment]
Hey @andrewjtait,

I created a PR #1840 for you to take a look at.  As I said before, Date Component will need unit tests.  Also, the red highlight for the input field of a date component has a tool tip that will always say invalid format.  For this particular case it should say "date out of range" or something.  Not sure where to implement that.

**Comment by ianforster-sage** [comment]
Carbon's form and validation functionality will be replaced with Formik 2.0.
https://github.com/Sage/carbon/blob/master/rfcs/text/formik.md
Our recommendation is to use Yup to validate your inputs.
